cedar-policy-validator/src/type_error.rs

```diff
@@ -331,9 +337,33 @@ pub(crate) enum UnexpectedTypeHelp {
 }
 
 /// Structure containing details about an incompatible type error.
-#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+#[derive(Diagnostic, Error, Debug, Clone, Hash, Eq, PartialEq)]
 pub struct IncompatibleTypes {
     pub(crate) types: BTreeSet<Type>,
+    #[help]
+    pub(crate) hint: LubHelp,
+}
+
+impl Display for IncompatibleTypes {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "The types ")?;
+        join_with_conjunction(f, "and", self.types.iter(), |f, t| write!(f, "{t}"))?;
+        write!(f, " are not compatible")
+    }
+}
+
+#[derive(Error, Debug, Clone, Hash, Eq, PartialEq)]
+pub(crate) enum LubHelp {
+    #[error("corresponding attributes of compatible record types must either both be required or both be optional")]
```

Comment:
Calling for more English experts. But I'd say "must be pairwise required or optional".